File: Mochila/mochila.js
//https://www.ibm.com/developerworks/community/blogs/hazem/entry/javascript_getting_all_possible_permutations?lang=en
var Util = {};

Util.getPermuts = function(array, start, output) {
    if(start >= array.length) {
        var arr = array.slice(0); //clone the array		
        output.push(arr);
    }
    else {
        var i;

        for(i = start; i < array.length; ++i) {
            Util.swap(array, start, i);
            Util.getPermuts(array, start + 1, output);
            Util.swap(array, start, i);
        }
    }
};

Util.getAllPossiblePermuts = function(array, output) {
    Util.getPermuts(array, 0, output);
};

Util.swap = function(array, from, to) {
    var tmp = array[from];
    array[from] = array[to];
    array[to] = tmp;
};

//setup
const bagCap = 100;
const itensCount = 5;
const maxItemWeight = 40;
const maxItemValue = 50;

var itens = new Array(itensCount);
var ids = new Array(itensCount);
var permutations = [];
var bestResult = 0;

for(var i = 0; i < itensCount; i++) {
    itens[i] = {};
    itens[i].value = Math.round(Math.random() * maxItemValue);
    itens[i].weight = Math.round(Math.random() * maxItemWeight);
    ids[i] = i;
}

//console.log(ids);
Util.getAllPossiblePermuts(ids, permutations);
//console.log(permutations);

console.time('mochilaTimer');
for(i = 0; i < permutations.length; i++) {
    let currentCap = 0;
    let totalVal = 0;

    for(var j = 0; j < itensCount; j++) {
        if(currentCap + itens[permutations[i][j]].weight <= bagCap) {
            currentCap += itens[permutations[i][j]].weight;
            totalVal += itens[permutations[i][j]].value;
        }
    }

    if(totalVal > bestResult)
        bestResult = totalVal;
}
console.timeEnd('mochilaTimer');

//console.log(bestResult);

File: Mochila/Mochila.java
package mochila;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class Mochila {
    public static class Item {
	int value, weight;

	public Item(int value, int weight) {
	    this.value = value;
	    this.weight = weight;
	}
    }

    static List<int[]> permutations = new ArrayList<>();
    static int[][] permutationsArray;

    public static void main(String[] args) {
	//setup
	final int bagCap = 150;
	final int itensCount = 8;
	final int maxItemWeight = 40;
	final int maxItemValue = 50;

	Item itens[] = new Item[itensCount];
	int ids[] = new int[itensCount];
	int bestResult = 0;

	for(int i = 0; i < itensCount; i++) {
	    itens[i] = new Item(Math.round((float) Math.random() * maxItemValue),
		    Math.round((float) Math.random() * maxItemWeight));
	    ids[i] = i;
	}
	permutationsArray = new int[factorial(itensCount)][itensCount];

	permute(itensCount, ids);
	//System.out.println(Arrays.deepToString(permutations.toArray(permutationsArray)));

	long startTime = System.nanoTime();
	for(int i = 0; i < permutationsArray.length; i++) {
	    int currentCap = 0;
	    int totalVal = 0;

	    for(int j = 0; j < itensCount; j++) {
		if(currentCap + itens[permutationsArray[i][j]].weight <= bagCap) {
		    currentCap += itens[permutationsArray[i][j]].weight;
		    totalVal += itens[permutationsArray[i][j]].value;
		}
	    }

	    if(totalVal > bestResult) {
		bestResult = totalVal;
	    }
	}
	long endTime = System.nanoTime();

	System.out.println("bubbleSortTimer: " + (endTime - startTime) / 1_000_000 + "ms");
    }

    public static void permute(int n, int[] elements) {

	if(n == 1) {
	    permutations.add(elements.clone());
	}
	else {
	    for(int i = 0; i < n - 1; i++) {
		permute(n - 1, elements);
		if(n % 2 == 0) {
		    swap(elements, i, n - 1);
		}
		else {
		    swap(elements, 0, n - 1);
		}
	    }
	    permute(n - 1, elements);
	}
    }

    private static void swap(int[] input, int a, int b) {
	int tmp = input[a];
	input[a] = input[b];
	input[b] = tmp;
    }

    public static int factorial(int value) {
	if(value == 1) {
	    return value;
	}
	else {
	    return value * factorial(value - 1);
	}
    }
}
